src/or/connection.c

```diff
@@ -2658,16 +2687,27 @@ retry_listener_ports(smartlist_t *old_conns,
           break;
         }
       } else {
-        int port_matches;
-        if (wanted->port == CFG_AUTO_PORT) {
-          port_matches = 1;
-        } else {
-          port_matches = (wanted->port == conn->port);
-        }
+        int port_matches_exact = (wanted->port == conn->port);
+        int port_matches = wanted->port || port_matches_exact;
+
         if (port_matches && tor_addr_eq(&wanted->addr, &conn->addr)) {
           found_port = wanted;
           break;
         }
+
+        int may_need_rebind = port_matches_exact &
+                              (tor_addr_is_null(&wanted->addr) ^
+                               tor_addr_is_null(&conn->addr));
+        if (replacements && may_need_rebind) {
+          struct replacement_s *replacement =
+            tor_malloc(sizeof(struct replacement_s));
+
+          replacement->old_conn = conn;
+          replacement->new_port = (port_cfg_t *)wanted;
```

Comment:
Instead of casting here, it would be safer to make the struct member into a const pointer.